File: Problems & Solutions/U/Unique partitions/21 November 2022/Python/Unique partitions.py
#User function Template for python3

class Solution:
	def UniquePartitions(self, n):
        # Code here
        
        def partitions(num,Sum,res,out):
            
            if Sum>target or num<=0:
                return 
            
            if Sum==target:
                out.append(res)
                return
            
            res1 = res[:]
            res2 = res[:]
            res1 = res1+[num]
            partitions(num,Sum+num,res1,out)
            partitions(num-1,Sum,res2,out)
            
            return
        
        Sum = 0
        res = []
        out = []
        target = n
        partitions(n,Sum,res,out)
        return out

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
	T=int(input())
	for i in range(T):
		n= int(input())
		ob = Solution()
		ans = ob.UniquePartitions(n)
		for a in ans:
			for b in a:
				print(b, end = " ")
		print()

# } Driver Code Ends

File: Problems & Solutions/U/Unique partitions/21 November 2022/C++/Unique partitions.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution{
	public:
	vector<vector<int>>ans;
    bool fun(int target,vector<int>&candidates,vector<int>res,int pos){
        if(target==0){
        	sort(res.begin(), res.end());
        	reverse(res.begin(), res.end());
            ans.push_back(res);
            return true;
        }
        if(target<0)
            return false;
        bool f=false;
        for(int i=pos;i<candidates.size();i++){
            res.push_back(candidates[i]);
            f=fun(target-candidates[i],candidates,res,i);
            res.pop_back();
        }
        return f;
    }    
    vector<vector<int>> UniquePartitions(int n) {
    	vector<int>cur, nums;
    	for(int i = 1; i <= n; i++)
    		nums.push_back(i);
    	fun(n, nums, cur, 0);
    	sort(ans.begin(), ans.end());
    	reverse(ans.begin(), ans.end());
    	return ans;
    }
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		Solution ob;
		vector<vector<int>> ans = ob.UniquePartitions(n);
    	for(auto i : ans)
    		for(auto j: i)
    			cout << j <<" ";
    	cout << "\n";
	}  
	return 0;
}
// } Driver Code Ends